File: js/Spreadsheet/AnnotationTable/ArcTable.js
import { iterateIndexed, length, sortBy, collect as collect_1, exists, toArray, ofArray, empty, tail, head, isEmpty, cons, singleton, fold, map, reverse } from "../../fable_modules/fable-library-js.4.22.0/List.js";
import { item, map as map_1 } from "../../fable_modules/fable-library-js.4.22.0/Array.js";
import { fold as fold_1, singleton as singleton_1, collect, map as map_2, delay, toArray as toArray_1, tryFind } from "../../fable_modules/fable-library-js.4.22.0/Seq.js";
import { toStringCellColumns, fixDeprecatedIOHeader, fromStringCellColumns } from "./CompositeColumn.js";
import { printf, toFail } from "../../fable_modules/fable-library-js.4.22.0/String.js";
import { Dictionary_tryGet, FsCellsCollection__TryGetCell_Z37302880 } from "../../fable_modules/FsSpreadsheet.6.3.0-alpha.4/Cells/FsCellsCollection.fs.js";
import { rangeDouble } from "../../fable_modules/fable-library-js.4.22.0/Range.js";
import { FsAddress_$ctor_Z37302880, FsAddress__get_ColumnNumber, FsAddress__get_RowNumber } from "../../fable_modules/FsSpreadsheet.6.3.0-alpha.4/FsAddress.fs.js";
import { FsRangeAddress_$ctor_7E77A4A0, FsRangeAddress__get_LastAddress } from "../../fable_modules/FsSpreadsheet.6.3.0-alpha.4/Ranges/FsRangeAddress.fs.js";
import { FsRangeBase__Cell_Z3407A44B, FsRangeBase__get_RangeAddress } from "../../fable_modules/FsSpreadsheet.6.3.0-alpha.4/Ranges/FsRangeBase.fs.js";
import { ArcTable } from "../../Core/Table/ArcTable.js";
import { FsWorksheet } from "../../fable_modules/FsSpreadsheet.6.3.0-alpha.4/FsWorksheet.fs.js";
import { comparePrimitives } from "../../fable_modules/fable-library-js.4.22.0/Util.js";
import { max } from "../../fable_modules/fable-library-js.4.22.0/Double.js";
import { addToDict } from "../../fable_modules/fable-library-js.4.22.0/MapUtil.js";

/**
 * Iterates over elements of the input list and groups adjacent elements.
 * A new group is started when the specified predicate holds about the element
 * of the list (and at the beginning of the iteration).
 * 
 * For example:
 * List.groupWhen isOdd [3;3;2;4;1;2] = [[3]; [3; 2; 4]; [1; 2]]
 */
export function Aux_List_groupWhen(f, list) {
    return reverse(map(reverse, fold((acc, e) => {
        const matchValue = f(e);
        if (matchValue) {
            return cons(singleton(e), acc);
        }
        else if (!isEmpty(acc)) {
            return cons(cons(e, head(acc)), tail(acc));
        }
        else {
            return singleton(singleton(e));
        }
    }, empty(), list)));
}

export function classifyHeaderOrder(header) {
    switch (header.tag) {
        case 4:
        case 5:
        case 6:
        case 7:
        case 8:
        case 9:
        case 10:
            return 2;
        case 0:
        case 1:
        case 2:
        case 3:
        case 14:
        case 13:
            return 3;
        case 12:
            return 4;
        default:
            return 1;
    }
}

export function classifyColumnOrder(column) {
    return classifyHeaderOrder(column.Header);
}

export const helperColumnStrings = ofArray(["Term Source REF", "Term Accession Number", "Unit", "Data Format", "Data Selector Format"]);

export function groupColumnsByHeader(stringCellColumns) {
    return map_1(toArray, toArray(Aux_List_groupWhen((c) => {
        const v = item(0, c);
        return !exists((s) => v.startsWith(s), helperColumnStrings);
    }, ofArray(stringCellColumns))));
}

/**
 * Returns the annotation table of the worksheet if it exists, else returns None
 */
export function tryAnnotationTable(sheet) {
    return tryFind((t) => t.Name.startsWith("annotationTable"), sheet.Tables);
}

/**
 * Groups and parses a collection of single columns into the according ISA composite columns
 */
export function composeColumns(stringCellColumns) {
    return map_1(fromStringCellColumns, groupColumnsByHeader(stringCellColumns));
}

/**
 * Returns the protocol described by the headers and a function for parsing the values of the matrix to the processes of this protocol
 */
export function tryFromFsWorksheet(sheet) {
    try {
        const matchValue = tryAnnotationTable(sheet);
        if (matchValue == null) {
            return undefined;
        }
        else {
            const t = matchValue;
            const compositeColumns = composeColumns(map_1(fixDeprecatedIOHeader, toArray_1(delay(() => map_2((c) => toArray_1(delay(() => collect((r) => {
                const matchValue_1 = FsCellsCollection__TryGetCell_Z37302880(sheet.CellCollection, r, c);
                if (matchValue_1 == null) {
                    return singleton_1("");
                }
                else {
                    const cell = matchValue_1;
                    return singleton_1(cell.ValueAsString());
                }
            }, rangeDouble(1, 1, FsAddress__get_RowNumber(FsRangeAddress__get_LastAddress(FsRangeBase__get_RangeAddress(t))))))), rangeDouble(1, 1, FsAddress__get_ColumnNumber(FsRangeAddress__get_LastAddress(FsRangeBase__get_RangeAddress(t)))))))));
            return ArcTable.addColumns(compositeColumns, undefined, true)(ArcTable.init(sheet.Name));
        }
    }
    catch (err) {
        const arg = sheet.Name;
        const arg_1 = err.message;
        return toFail(printf("Could not parse table with name \"%s\":\n%s"))(arg)(arg_1);
    }
}

export function toFsWorksheet(index, table) {
    const stringCount = new Map([]);
    const ws = new FsWorksheet(table.Name);
    if (table.ColumnCount === 0) {
        return ws;
    }
    else {
        const columns = collect_1(toStringCellColumns, sortBy(classifyColumnOrder, ofArray(table.Columns), {
            Compare: comparePrimitives,
        }));
        let tableRowCount;
        const maxRow = fold_1((acc, c) => max(acc, length(c)), 0, columns) | 0;
        tableRowCount = ((maxRow === 1) ? 2 : maxRow);
        const tableColumnCount = length(columns) | 0;
        const name = (index == null) ? "annotationTable" : (`${"annotationTable"}${index}`);
        const fsTable = ws.Table(name, FsRangeAddress_$ctor_7E77A4A0(FsAddress_$ctor_Z37302880(1, 1), FsAddress_$ctor_Z37302880(tableRowCount, tableColumnCount)));
        iterateIndexed((colI, col) => {
            iterateIndexed((rowI, stringCell) => {
                let value;
                if (rowI === 0) {
                    const matchValue = Dictionary_tryGet(stringCell, stringCount);
                    if (matchValue == null) {
                        addToDict(stringCount, stringCell, "");
                        value = stringCell;
                    }
                    else {
                        const spaces = matchValue;
                        stringCount.set(stringCell, spaces + " ");
                        value = ((stringCell + " ") + spaces);
                    }
                }
                else {
                    value = stringCell;
                }
                const address = FsAddress_$ctor_Z37302880(rowI + 1, colI + 1);
                FsRangeBase__Cell_Z3407A44B(fsTable, address, ws.CellCollection).SetValueAs(value);
            }, col);
        }, columns);
        return ws;
    }
}

File: src/Spreadsheet/AnnotationTable/ArcTable.fs
module ARCtrl.Spreadsheet.ArcTable

open ARCtrl
open ARCtrl.Helper
open FsSpreadsheet

// I think we really should not add FSharpAux for exactly one function.
module Aux =

    module List =

        /// Iterates over elements of the input list and groups adjacent elements.
        /// A new group is started when the specified predicate holds about the element
        /// of the list (and at the beginning of the iteration).
        ///
        /// For example: 
        ///    List.groupWhen isOdd [3;3;2;4;1;2] = [[3]; [3; 2; 4]; [1; 2]]
        let groupWhen f list =
            list
            |> List.fold (
                fun acc e ->
                    match f e, acc with
                    | true  , _         -> [e] :: acc       // true case
                    | false , h :: t    -> (e :: h) :: t    // false case, non-empty acc list
                    | false , _         -> [[e]]            // false case, empty acc list
            ) []
            |> List.map List.rev
            |> List.rev
 

type ColumnOrder =
    | InputClass = 1
    | ProtocolClass = 2
    | ParamsClass = 3
    | OutputClass = 4

let classifyHeaderOrder (header : CompositeHeader) =     
    match header with
    | CompositeHeader.Input             _ -> ColumnOrder.InputClass

    | CompositeHeader.ProtocolType          
    | CompositeHeader.ProtocolDescription
    | CompositeHeader.ProtocolUri
    | CompositeHeader.ProtocolVersion
    | CompositeHeader.ProtocolREF       
    | CompositeHeader.Performer
    | CompositeHeader.Date                -> ColumnOrder.ProtocolClass

    | CompositeHeader.Component         _
    | CompositeHeader.Characteristic    _
    | CompositeHeader.Factor            _
    | CompositeHeader.Parameter         _ 
    | CompositeHeader.Comment           _
    | CompositeHeader.FreeText          _ -> ColumnOrder.ParamsClass

    | CompositeHeader.Output            _ -> ColumnOrder.OutputClass

let classifyColumnOrder (column : CompositeColumn) =
    column.Header
    |> classifyHeaderOrder

[<Literal>]
let annotationTablePrefix = "annotationTable"

let helperColumnStrings = 
    [
        "Term Source REF"
        "Term Accession Number"
        "Unit"
        "Data Format"
        "Data Selector Format"
    ]

let groupColumnsByHeader (stringCellColumns : array<string []>) = 
    stringCellColumns
    |> Array.toList
    |> Aux.List.groupWhen (fun c -> 
        let v = c.[0]
        helperColumnStrings
        |> List.exists (fun s -> v.StartsWith s) 
        |> not
    )
    |> Array.ofList
    |> Array.map Array.ofList

/// Returns the annotation table of the worksheet if it exists, else returns None
let tryAnnotationTable (sheet : FsWorksheet) =
    sheet.Tables
    |> Seq.tryFind (fun t -> t.Name.StartsWith annotationTablePrefix)

/// Groups and parses a collection of single columns into the according ISA composite columns
let composeColumns (stringCellColumns : array<string []>) : CompositeColumn [] =
    stringCellColumns
    |> groupColumnsByHeader
    |> Array.map CompositeColumn.fromStringCellColumns

/// Returns the protocol described by the headers and a function for parsing the values of the matrix to the processes of this protocol
let tryFromFsWorksheet (sheet : FsWorksheet) =
    try
        match tryAnnotationTable sheet with
        | Some (t: FsTable) -> 
            let stringCellColumns = 
                [|
                for c = 1 to t.RangeAddress.LastAddress.ColumnNumber do
                    [|for r = 1 to t.RangeAddress.LastAddress.RowNumber do
                        match sheet.CellCollection.TryGetCell(r,c) with
                        | Some cell -> cell.ValueAsString()
                        | None -> ""
                    |]
                |]              
            let compositeColumns = 
                stringCellColumns
                |> Array.map CompositeColumn.fixDeprecatedIOHeader
                |> composeColumns
            ArcTable.init sheet.Name
            |> ArcTable.addColumns(compositeColumns)
            |> Some
        | None ->
            None
    with
    | err -> failwithf "Could not parse table with name \"%s\":\n%s" sheet.Name err.Message


let toFsWorksheet (index : int option) (table : ArcTable) =
    /// This dictionary is used to add spaces at the end of duplicate headers.
    let stringCount = System.Collections.Generic.Dictionary<string,string>()
    let ws = FsWorksheet(table.Name)

    // Cancel if there are no columns
    if table.ColumnCount = 0 then ws
    else

    let columns = 
        table.Columns
        |> List.ofSeq
        |> List.sortBy classifyColumnOrder
        |> List.collect CompositeColumn.toStringCellColumns
    let tableRowCount =
        let maxRow = columns |> Seq.fold (fun acc c -> max acc c.Length) 0
        if maxRow = 1 then 2
        else maxRow
    let tableColumnCount = columns.Length
    let name =
        match index with
        | Some i -> $"{annotationTablePrefix}{i}"
        | None -> annotationTablePrefix
    let fsTable = ws.Table(name,FsRangeAddress(FsAddress(1,1),FsAddress(tableRowCount,tableColumnCount)))
    columns
    |> List.iteri (fun colI col ->         
        col
        |> List.iteri (fun rowI stringCell -> 
            let value = 
                if rowI = 0 then
                    
                    match Dictionary.tryGet stringCell stringCount with
                    | Some spaces ->
                        stringCount.[stringCell] <- spaces + " "
                        stringCell + " " + spaces
                    | None ->
                        stringCount.Add(stringCell,"")
                        stringCell
                else stringCell
            let address = FsAddress(rowI+1,colI+1)
            fsTable.Cell(address, ws.CellCollection).SetValueAs value
        )  
    )
    ws